homeassistant/components/slide/__init__.py

```diff
@@ -0,0 +1,156 @@
+        else:
+            _LOGGER.warning("Slide API returned 0 slides")
+
+        for slide in result:
+            if "device_id" not in slide:
+                _LOGGER.error(
+                    "Found invalid Slide entry, device_id is " "missing. Entry=%s",
+                    slide,
+                )
+                continue
+
+            uid = slide["device_id"].replace("slide_", "")
+            slidenew = hass.data[DOMAIN][SLIDES].setdefault(uid, {})
+            slidenew["mac"] = uid
+            slidenew["id"] = slide["id"]
+            slidenew["name"] = slide["device_name"]
+            slidenew["state"] = None
+            oldpos = slidenew.get("pos")
+            slidenew["pos"] = None
+            slidenew["online"] = False
+
+            if "device_info" not in slide:
+                _LOGGER.error(
+                    "Slide %s (%s) has no device_info Entry=%s",
+                    slide["id"],
+                    slidenew["mac"],
+                    slide,
+                )
+                continue
+
+            # Check if we have pos (OK) or code (NOK)
+            if "pos" in slide["device_info"]:
+                slidenew["online"] = True
+                slidenew["pos"] = slide["device_info"]["pos"]
+                slidenew["pos"] = max(0, min(1, slidenew["pos"]))
+
+                if oldpos is None or oldpos == slidenew["pos"]:
+                    slidenew["state"] = (
+                        STATE_CLOSED if slidenew["pos"] > 0.95 else STATE_OPEN
+                    )
+                elif oldpos < slidenew["pos"]:
+                    slidenew["state"] = (
+                        STATE_CLOSED if slidenew["pos"] >= 0.95 else STATE_CLOSING
+                    )
+                else:
+                    slidenew["state"] = (
+                        STATE_OPEN if slidenew["pos"] <= 0.05 else STATE_OPENING
+                    )
+            elif "code" in slide["device_info"]:
+                _LOGGER.warning(
+                    "Slide %s (%s) is offline with " "code=%s",
+                    slide["id"],
+                    slidenew["mac"],
+                    slide["device_info"]["code"],
+                )
+            else:
+                _LOGGER.error(
+                    "Slide %s (%s) has invalid device_info %s",
+                    slide["id"],
+                    slidenew["mac"],
+                    slide["device_info"],
+                )
+
+            _LOGGER.debug("Updated entry=%s", slidenew)
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][SLIDES] = {}
+
+    username = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    scaninterval = config[DOMAIN][CONF_SCAN_INTERVAL]
+
+    hass.data[DOMAIN][API] = GoSlideCloud(username, password)
+
+    # pylint: disable=broad-except
+    try:
+        result = await hass.data[DOMAIN][API].login()
+    except (ClientConnectionError, ClientConnectorError, TimeoutError) as err:
```

Comment:
It's better if the library raises library specific exceptions, with a common library specific parent exception.